Support duplicate IDs in Config

MatthewMuehlhauserNRCan opened the pull request:
# Description

This change allows there to be multiple layers with the same ID. Specifically, this is meant for GeoCore layers where a user may want to split the sub-features of the GeoCore layer between two or more layers. This is done by adding a modifier to the end of the UUID using a colon and an index (e.g. ':2'). When fetching for the GeoCore layer, anything including and after the ':' is ignored so the GeoCore fetch will work properly. This also means, instead of just an index number, you could use a short description as well (e.g. ':water') depending on the use case.

That said, this will affect all layers which may have duplicate layer IDs and not just GeoCore.

Fixes # 2771 - Create many Geoview layers from GeoCore with same UUID

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
### Host Updated 2025-04-02 9:20am

Navigator page has two different sets of layers. Two that share the same UUID and two that share the UUID but one has a modified UUID with ":2" to make it unique.

Also tested adding same geocore id twice through the "Add Layer" in the layers tab. (5a65ad7c-561a-466a-8375-8d876624df9d)

[Navigator - Layer GeoView - Duplicate GeoCore Config](https://matthewmuehlhausernrcan.github.io/geoview/demos-navigator.html?config=./configs/navigator/28-geocore-duplicates.json)

[OSDP Air Demo](https://matthewmuehlhausernrcan.github.io/geoview/demo-osdp-air.html)

# Checklist:

- [X] I have build __(rush build)__ and deploy __(rush host)__ my PR
- [X] I have connected the issues(s) to this PR
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] I have created new issue(s) related to the outcome of this PR is needed
-  ~~I have made corresponding changes to the documentation~~
-  ~~I have added tests that prove my fix is effective or that my feature works~~
-  ~~New and existing unit tests pass locally with my changes~~

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)
<!-- Reviewable:end -->

MatthewMuehlhauserNRCan committed:
Support duplicate IDs in Config

MatthewMuehlhauserNRCan committed:
Update GeoCore Geocharts for duplicate Ids

MatthewMuehlhauserNRCan committed:
Update OSDP Air config

MatthewMuehlhauserNRCan committed:
Handle duplicate GeoCore layers being added through the layer tab

MatthewMuehlhauserNRCan committed:
Handle duplicates in addGeoviewLayerByGeoCoreUUID

jolevesq reviewed (changes requested):
> Reviewed 3 of 8 files at r1, 3 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: 6 of 10 files reviewed, 3 unresolved discussions (waiting on @MatthewMuehlhauserNRCan)

___
*[`packages/geoview-core/public/templates/demos/demo-osdp-integration.html` line 327 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS7n3M0z9OtAwSA5NB:-OMS7n3ND748xKzASrzo:b5lafif) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/cf7cf1ef2a7544fdc5e2e8c4ec39f4d086ef201a/packages/geoview-core/public/templates/demos/demo-osdp-integration.html#L327)):*
> ```HTML
>       function addLayers(layers) {
>         layers.forEach((layer, i) => {
>           if (layers.indexOf(layer) < i) {
> ```

I think this should handle in addGeoviewLayerByGeocoreUUID. If it exist, happen the custom section. The first one to go in is the real uuid (link with geochart if needed). The others are uuid:custom
___
*[`packages/geoview-core/src/geo/layer/layer.ts` line 652 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8mPjEKEJg7uiIhM9:-OMS8mPkCPlD1bC-qIvY:btbdtxi) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/4b88702a6e600643f6f28b67d365df00a233e587/packages/geoview-core/src/geo/layer/layer.ts#L652)):*
> ```XML
> 
>     // TODO: Refactor - This should be dealt with the config classes and this line commented out, therefore, content of addGeoviewLayerStep2 becomes this addGeoviewLayer function.
>     if (geoviewLayerConfig.geoviewLayerId in this.#geoviewLayers) {
> ```

Do we want to remove all duplicate or oonly for geocore. Because if a user add 2 layer by config they should not have the same id. If so we happen thevalue but if he use it elsewere the id he enter is now wrong. We can let this happen and add the custom part but we should add a notification so we know something is not good
___
*[`packages/geoview-core/src/core/components/layers/left-panel/add-new-layer/add-new-layer.tsx` line 430 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8IJUEn3QqH8FruGJ:-OMS8IJUEn3QqH8FruGK:b-z7tzoa) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/cf7cf1ef2a7544fdc5e2e8c4ec39f4d086ef201a/packages/geoview-core/src/core/components/layers/left-panel/add-new-layer/add-new-layer.tsx#L430)):*
> ```XML
> 
>       const geoCoreGeoviewLayerInstance = new GeoCore(mapId, api.maps[mapId].getDisplayLanguage());
>       // Add uniqueId to all geoCore layers so they will never be a duplicate layerPath
> ```

The happen of custom uuid should be handle in createLAyersFrom uuid I think


<!-- Sent from Reviewable.io -->

MatthewMuehlhauserNRCan committed:
Handle duplicates in GeoCore and other cleanup

MatthewMuehlhauserNRCan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: 5 of 11 files reviewed, 3 unresolved discussions (waiting on @jolevesq and @MatthewMuehlhauserNRCan)

___
*[`packages/geoview-core/public/templates/demos/demo-osdp-integration.html` line 327 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS7n3M0z9OtAwSA5NB:-OMSGZltAk12DIMpyzQr:bh74dfu) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/cf7cf1ef2a7544fdc5e2e8c4ec39f4d086ef201a/packages/geoview-core/public/templates/demos/demo-osdp-integration.html#L327)):*
<details><summary><i>Previously, jolevesq (Johann Levesque) wrote…</i></summary><blockquote>

I think this should handle in addGeoviewLayerByGeocoreUUID. If it exist, happen the custom section. The first one to go in is the real uuid (link with geochart if needed). The others are uuid:custom
</blockquote></details>

So, this is mostly handled in the addGeoviewLayerByGeocoreUUID, however; in the case where there are two of the same UUID in the input textbox, then both layers are going through the addGeoviewLayerByGeocoreUUID process at the same time and won't be aware of the other. This is just to catch that case.
___
*[`packages/geoview-core/src/core/components/layers/left-panel/add-new-layer/add-new-layer.tsx` line 430 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8IJUEn3QqH8FruGJ:-OMTDH5MBnVxIBnsSbsJ:bb6un1j) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/cf7cf1ef2a7544fdc5e2e8c4ec39f4d086ef201a/packages/geoview-core/src/core/components/layers/left-panel/add-new-layer/add-new-layer.tsx#L430)):*
<details><summary><i>Previously, jolevesq (Johann Levesque) wrote…</i></summary><blockquote>

The happen of custom uuid should be handle in createLAyersFrom uuid I think
</blockquote></details>

Done
___
*[`packages/geoview-core/src/geo/layer/layer.ts` line 652 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8mPjEKEJg7uiIhM9:-OMTDL1lCd-X0ycXHqGK:b-xafbc7) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/4b88702a6e600643f6f28b67d365df00a233e587/packages/geoview-core/src/geo/layer/layer.ts#L652)):*
<details><summary><i>Previously, jolevesq (Johann Levesque) wrote…</i></summary><blockquote>

Do we want to remove all duplicate or oonly for geocore. Because if a user add 2 layer by config they should not have the same id. If so we happen thevalue but if he use it elsewere the id he enter is now wrong. We can let this happen and add the custom part but we should add a notification so we know something is not good
</blockquote></details>

Added back in


<!-- Sent from Reviewable.io -->

MatthewMuehlhauserNRCan committed:
Remove unnecessary comment

MatthewMuehlhauserNRCan commented:
*[`packages/geoview-core/src/geo/layer/layer.ts` line 652 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8mPjEKEJg7uiIhM9:-OMgtC9TC1q6iU6mS3uY:b29qz6l) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/4b88702a6e600643f6f28b67d365df00a233e587/packages/geoview-core/src/geo/layer/layer.ts#L652)):*
<details><summary><i>Previously, MatthewMuehlhauserNRCan wrote…</i></summary><blockquote>

Added back in
</blockquote></details>

So, I added this back in, however; in the utils > config > config.ts, a duplicate marker is added to any geoviewId, regardless if it's geocore or not, so this isn't actually called anyway. Do we want to make this exclusive to GeoCore only so other layers do have to be made unique by the user?


<!-- Sent from Reviewable.io -->

jolevesq reviewed (changes requested):
> Reviewed 1 of 6 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: all files reviewed, 1 unresolved discussion (waiting on @MatthewMuehlhauserNRCan)

___
*[`packages/geoview-core/public/templates/demos/demo-osdp-integration.html` line 327 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS7n3M0z9OtAwSA5NB:-OMlGvz58Pzr3wBYAyK3:bd1nqyo) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/cf7cf1ef2a7544fdc5e2e8c4ec39f4d086ef201a/packages/geoview-core/public/templates/demos/demo-osdp-integration.html#L327)):*
<details><summary><i>Previously, MatthewMuehlhauserNRCan wrote…</i></summary><blockquote>

So, this is mostly handled in the addGeoviewLayerByGeocoreUUID, however; in the case where there are two of the same UUID in the input textbox, then both layers are going through the addGeoviewLayerByGeocoreUUID process at the same time and won't be aware of the other. This is just to catch that case.
</blockquote></details>

Oh, I see now
___
*[`packages/geoview-core/src/geo/layer/layer.ts` line 652 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8mPjEKEJg7uiIhM9:-OMlGNW88AGzkHT6z9Qp:b-7ja9iq) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/4b88702a6e600643f6f28b67d365df00a233e587/packages/geoview-core/src/geo/layer/layer.ts#L652)):*
<details><summary><i>Previously, MatthewMuehlhauserNRCan wrote…</i></summary><blockquote>

So, I added this back in, however; in the utils > config > config.ts, a duplicate marker is added to any geoviewId, regardless if it's geocore or not, so this isn't actually called anyway. Do we want to make this exclusive to GeoCore only so other layers do have to be made unique by the user?
</blockquote></details>

I would say yes we need to do this only for geocore or maybe instead of throwing an error, we just throw a warning. If not on pupose, duplicate id may have side effect...


<!-- Sent from Reviewable.io -->

Alex-NRCan reviewed (changes requested):
> Reviewed 3 of 8 files at r1, 3 of 6 files at r2, 4 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: all files reviewed, 2 unresolved discussions (waiting on @MatthewMuehlhauserNRCan)

___
*[`packages/geoview-core/src/geo/map/map-viewer.ts` line 665 at r3](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMlUW0x1fI8I336E2_q:-OMlUW0x1fI8I336E2_r:b-j8gn5z) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/421c7fb9c4b64e533fc817bb2a3fac9bebc8cd18/packages/geoview-core/src/geo/map/map-viewer.ts#L665)):*
> ```XML
> 
>       this.onMapLayersLoaded(() => {
>         let layerExtents = this.layer.getExtentOfMultipleLayers(layerIdsToZoomTo);
> ```

Here we're sending layer paths instead of layer ids to getExtentOfMultipleLayers which still expects layerIds (or layerIds OR layerpaths? unsure) in its param.
If that's on purpose, maybe update the getExtentOfMultipleLayers function and update its comments too?


<!-- Sent from Reviewable.io -->

MatthewMuehlhauserNRCan committed:
Re-add duplicate error for user provided IDs and update duplicate config for swiper plugin

MatthewMuehlhauserNRCan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: all files reviewed, 2 unresolved discussions (waiting on @Alex-NRCan and @jolevesq)

___
*[`packages/geoview-core/src/geo/layer/layer.ts` line 652 at r2](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMS8mPjEKEJg7uiIhM9:-OMluWWb7hD6O6BR0X2V:bb6un1j) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/4b88702a6e600643f6f28b67d365df00a233e587/packages/geoview-core/src/geo/layer/layer.ts#L652)):*
<details><summary><i>Previously, jolevesq (Johann Levesque) wrote…</i></summary><blockquote>

I would say yes we need to do this only for geocore or maybe instead of throwing an error, we just throw a warning. If not on pupose, duplicate id may have side effect...
</blockquote></details>

Done
___
*[`packages/geoview-core/src/geo/map/map-viewer.ts` line 665 at r3](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818#-OMlUW0x1fI8I336E2_q:-OMluhqcEtwKaRJ5kIMg:b-1csi4v) ([raw file](https://github.com/Canadian-Geospatial-Platform/geoview/blob/421c7fb9c4b64e533fc817bb2a3fac9bebc8cd18/packages/geoview-core/src/geo/map/map-viewer.ts#L665)):*
<details><summary><i>Previously, Alex-NRCan (Alex) wrote…</i></summary><blockquote>

Here we're sending layer paths instead of layer ids to getExtentOfMultipleLayers which still expects layerIds (or layerIds OR layerpaths? unsure) in its param.
If that's on purpose, maybe update the getExtentOfMultipleLayers function and update its comments too?
</blockquote></details>

Ah I see. The getGeoviewLayerIds was previously returning layerPaths (just tested for the sandbox map, gave me wmsLYR1/0 instead of wmsLYR1), which was the reason for the rename and the creation of the "new" getGeoviewLayerIds() as seen in layer.ts, so I should be able to just change it back. Looks like this worked before, but only because, in the layer.ts > getExtentOfMultipleLayers, the filter also checks if the layerPath is equal to the layerId in addition to checking if it startsWith the layerId.


<!-- Sent from Reviewable.io -->

DamonU2 reviewed (approved):
> Reviewed 2 of 8 files at r1, 2 of 6 files at r2, 3 of 5 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: all files reviewed, 2 unresolved discussions (waiting on @Alex-NRCan and @jolevesq)



<!-- Sent from Reviewable.io -->

MatthewMuehlhauserNRCan committed:
Change back to getGeoviewLayerIds

Alex-NRCan reviewed (approved):
> Reviewed 1 of 5 files at r3, 2 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: all files reviewed, 1 unresolved discussion (waiting on @jolevesq)



<!-- Sent from Reviewable.io -->

MatthewMuehlhauserNRCan committed:
Change random unique IDs to only use first 8 characters for readability

jolevesq reviewed (approved):
> Reviewed 4 of 4 files at r4, 1 of 1 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Canadian-Geospatial-Platform/geoview/2818)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @MatthewMuehlhauserNRCan)



<!-- Sent from Reviewable.io -->
